File: codingjs/count8.js
function count8(n: number): number {
  if (n < 8) return 0;
  return count8Helper(n, 0);
}

function count8Helper(n: number, count: number, isPrevEight: boolean): number {
  if (n === 0) return count;
  let eightCount = 0;
  if (n % 10 === 8) eightCount = 1;
  if (n % 100 === 88) eightCount = 2;
  return count8Helper(Math.floor(n / 10), count + eightCount);
}

File: codingbat/java/count8.java
public int count8Helper(int n, int count, boolean isPrevEight) {
  if (n < 10) {
    if (n == 8 && isPrevEight) return count + 2;
    else if (n == 8) return count + 1;
    else return count;
  }
  int remainder = n % 10;
  int quotient = n / 10;
  if (remainder == 8) {
    if (isPrevEight) count += 2;
    else count += 1;
    isPrevEight = true;
  } else {
    isPrevEight = false;
  }
  return count8Helper(quotient, count, isPrevEight);
}

public int count8(int n) {
  return count8Helper(n, 0, false);
}

